### DISCUSSION TITLE
.NET8 Blazor Auto Render Mode - How to prevent sending DataPortal code to client?

### DISCUSSION THREAD

**Discussion by MarkOverstreet** [MarimerLLC/csla#4158] [Questions]
Our goal is to keep the data access code off of the wasm client because we have an app that has ADO.NET code directly in the business object that inherits from BusinessBase.  In .NET6, we accomplished this by having our business object split between two files as a partial class.  The domain class contained the properties, business rules, etc. while the dataportal.cs class contained the Fetch, Insert, Update, etc. with ADO.NET.  When compiling the data portal project, it included both files; however, when compiling the WASM project, it ignored any file ending with "dataportal.cs".   This worked great since the client never actually uses the local data portal but instead calls into the remote data portal.

EX:  
Gift.cs
Gift.dataportal.cs

In .NET8, you build one project to publish which is the server project it handles building the all the referenced projects including the wasm project and now our previous approach does not work.  How are others accomplishing this currently?  Any suggestions on how to accomplish this?

Thanks
Mark

**Comment by kcabral817** [comment]
Good Question.  I do a similar thing with business object and a separate data access layer.


Kevin




________________________________
From: Mark Overstreet ***@***.***>
Sent: Monday, August 12, 2024 8:20 AM
To: MarimerLLC/csla ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [MarimerLLC/csla] .NET8 Blazor Auto Render Mode - How to prevent sending DataPortal code to client? (Discussion #4158)


Our goal is to keep the data access code off of the wasm client because we have an app that has ADO.NET code directly in the business object that inherits from BusinessBase. In .NET6, we accomplished this by having our business object split between two files as a partial class. The domain class contained the properties, business rules, etc. while the dataportal.cs class contained the Fetch, Insert, Update, etc. with ADO.NET. When compiling the data portal project, it included both files; however, when compiling the WASM project, it ignored any file ending with "dataportal.cs". This worked great since the client never actually uses the local data portal but instead calls into the remote data portal.

EX:
Gift.cs
Gift.dataportal.cs

In .NET8, you build one project to publish which is the server project it handles building the all the referenced projects including the wasm project and now our previous approach does not work. How are others accomplishing this currently? Any suggestions on how to accomplish this?

Thanks
Mark

**Comment by bboney** [comment]
Mark,

We do this by breaking out the ADO.NET code into a separate class library that implements a DAL interface.

So we have these projects:

- Project.Dal - Interface for DAL (No References)
- Project.DalSql - Implements interface with ADO.NET (References Csla.Data.SqlClient & Project.Dal)
- Project.Business - Business Objects (References Csla & Project.Dal)
- Project.Blazor - "Server" project in Global WebAssembly interactivity mode (References Project.Business, Project.DalSql, Project.Dal)
- Project.Blazor.Client - wasm project (References Project.Business & Project.Dal)

In your scenario, you may be able to use compiler directives in your code like [UnsupportedOSPlatform("browser")]. Csla uses this in the DataPortalClient code. We have not tried this so let us know what you find out.

`#if NET8_0_OR_GREATER
    [UnsupportedOSPlatform("browser")]
#endif
public class DalMethodsClass
{

}`

Hope this helps!

Ben

**Reply by MarkOverstreet** [reply]
Interesting idea. We may try that.

We could definitely use the interface/repository approach but sometimes, we have very little time to complete a project, and it was SO easy before to add our ADO.NET code directly in the partial class and just not include those in the build for wasm client!

**Reply by rockfordlhotka** [reply]
I generally do what @bboney suggests.
